### PR TITLE
User Contexts

### LINKED ISSUE #30: Migrate Localstorage to Contexts

### PR DESCRIPTION
**Pull request by curtisjbradley**
Context data is saved in native react contexts as opposed to localstorage. Now only tokens are saved.

Header updates depending on if the user is logged in.

Admin dashboard only appears if the user has permissions to use it.

### CODE REVIEW

File: frontend/src/other/UserContext.jsx
Comment (heyitsjorge) [inline frontend/src/other/UserContext.jsx:1]: This is a great addition, Curtis! It's really crucial for us to maintain a state for each kind of user!

File: frontend/App.jsx [hunk 1]
```diff
@@ -14,24 +14,54 @@ import NotFound from './src/pages/NotFound.jsx';
 import AdminDashboard from './src/pages/AdminDashboard.jsx';
 import ProductEdit from './src/pages/ProductEdit.jsx';
 import Cart from './src/pages/Cart.jsx';
+import { UserContext } from './src/other/UserContext.jsx';
+import { jwtDecode } from 'jwt-decode';
```
Comment (heyitsjorge) [inline frontend/App.jsx:18]: Cool JWT's for authentication!

File: frontend/src/pages/Cart.jsx [hunk 2]
```diff
@@ -1,25 +1,32 @@
...
     const update = (product, delta) => {
-        for (let i = 0; i < cart.length; i++) {
-            if (!cart[i]) continue;
-            if (cart[i]['itemId'] === product) {
-                cart[i].quantity += delta;
-                if (cart[i].quantity < 1) {
-                    cart.splice(i, 1);
-                }
-            }
+        const i = cart.findIndex((prod) => prod['itemId'] === product);
```
Comment (heyitsjorge) [inline frontend/src/pages/Cart.jsx:23]: Nice use of the findIndex method!

**Review by heyitsjorge — COMMENTED** [review]
A lot of this code was focused on implementing the JWT'S for authentication and allowing us to switch off the sites look/functionalty depending on the kind of user currently signed in. Great Job and Thankyou!

File: frontend/src/components/Header.jsx
Comment (phillip-niakoulov) [inline frontend/src/components/Header.jsx:1]: I spent like 2 hours wrangling Redirect to try and get this to work. Looks clean.

File: frontend/App.jsx (hunk 1 above)
Comment (edsonmunoz) [reply frontend/App.jsx:18]: Looks good!!!

File: frontend/src/pages/Login.jsx
Comment (edsonmunoz) [inline frontend/src/pages/Login.jsx:1]: good use of useContext to store data.

**Pull request merged**